### PR TITLE
Hyperparameters docs refresh

### LINKED ISSUE #9984: save_hyperparameters() saves all **kwargs arguments - which includes unpickable plugins for DDP
## 🐛 Bug

The call to save `self.save_hyperparameters()` (NOTE: no args provided) as described in the docs (see [here](https://pytorch-lightning.readthedocs.io/en/latest/common/hyperparameters.html#lightningmodule-hyperparameters
)) is by default saving the content of **kwargs to hparams. This is an issue because:

1. this is not matching the documentation (see the 'equivalent' call part)
2. one of the fields (I suspect introduced between 1.2.8 to 1.4.9?) is 'plugins' and there there is a DDP related object which cannot be pickled. This causes model checkpoints saved like this to be impossible to be loaded outside of a similar env (boh with PL loading functionality or the torch.load method).

### To Reproduce

Save any model according to the docs and try to load it in another script which runs afterwards (i.e. in a inference only script) not powered by PL

### Expected behavior

It should keep working as it did in version 1.2.8

### Environment

- PyTorch Lightning Version 1.4.9

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

A refresh of the docs around the `save_hyperparameters()` method. The current text is dated and does no longer accurately reflect the recommended, best possible ways to deal with hyperparameters. 
The `save_hyperparameters` method should not be seen as a magical feature. There is one way to use it and it must be cristal clear how to instantiate a model from a checkpoint.

1. Several examples had a "**kwargs" in their signature. While this can certainly be handled by our saving logic, it's not the first thing we want to recommend to users. It's a bad practice as unwanted objects could be accidentally getting pickled into the checkpoints, and make them potentially not able to load again.

2. It mentioned that one can save whole containers as `save_hyperparametes(conf)`. While this is possible, this should be avoided at all cost and is an artifact of a backward compatibility with the "old" way (pre 1.0) of saving hyperparameters. If one wants to save a  container of paramters (e.g. OmegaConf dict), then one should treat it like any other hyperaparameter and do `self.save_hyperparameters("conf")`. Otherwise, it causes too many problems, especially when the user does not know how to use it properly and later on finds that they can't reload their checkpoint with a simple `Model.load_from_checkpoint` as advertised in the docs. This has been a pain point for ages. Now, we will NO LONGER document this.

See all these issues from the past:
#3998, #4333, #4390, #5345

Fixes #9984
Should help with repeated questions asked on Slack

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

### CODE REVIEW

File: docs/source/common/hyperparameters.rst [hunk 1]
```diff
@@ -113,78 +113,75 @@ LightningModule hyperparameters
...
-                # save the config and any extra arguments
-                self.save_hyperparameters(conf)
-                self.save_hyperparameters(kwargs)
+load_from_checkpoint
+""""""""""""""""""""
 
-                self.layer_1 = nn.Linear(28 * 28, self.hparams.layer_1_dim)
-                self.layer_2 = nn.Linear(self.hparams.layer_1_dim, self.hparams.layer_2_dim)
-                self.layer_3 = nn.Linear(self.hparams.layer_2_dim, 10)
+LightningModule's that have hyperparameters automatically saved with :meth:`~pytorch_lightning.core.lightning.LightningModule.save_hyperparameters`
```
Comment (awaelchli) [inline docs/source/common/hyperparameters.rst:169]:
```suggestion
LightningModules that have hyperparameters automatically saved with :meth:`~pytorch_lightning.core.lightning.LightningModule.save_hyperparameters`
```

File: docs/source/common/hyperparameters.rst [hunk 2]
```diff
@@ -113,78 +113,75 @@ LightningModule hyperparameters
 Often times we train many versions of a model. You might share that model or come back to it a few months later
 at which point it is very useful to know how that model was trained (i.e.: what learning rate, neural network, etc...).
 
-Lightning has a few ways of saving that information for you in checkpoints and yaml files. The goal here is to
+Lightning has a standardized way saving that information for you in checkpoints and yaml files. The goal here is to
```
Comment (tchaton) [inline docs/source/common/hyperparameters.rst:116]:
```suggestion
Lightning has a standardized way of saving that information for you in checkpoints and YAML files. The goal here is to
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: docs/source/common/hyperparameters.rst [hunk 3]
```diff
@@ -113,78 +113,75 @@ LightningModule hyperparameters
...
+            self.save_hyperparameters(exclude=["loss_fx", "generator_network"])
 
-    .. code-block:: python
 
-        class LitMNIST(LightningModule):
-            def __init__(self, conf: Optional[Union[Dict, Namespace, DictConfig]] = None, **kwargs):
-                super().__init__()
-                # save the config and any extra arguments
-                self.save_hyperparameters(conf)
-                self.save_hyperparameters(kwargs)
```
Comment (carmocca) [inline docs/source/common/hyperparameters.rst:169]: I understand you don't want to document this as it can be a footgun for users.

But in that case, should also deprecate this behaviour? What's the point of having support but keep it undocumented?

Comment (awaelchli) [reply docs/source/common/hyperparameters.rst:169]: It is still documented here in the method signature: https://github.com/PyTorchLightning/pytorch-lightning/blob/a9bd4fbd96c1e73e251859d99b207008008de87d/pytorch_lightning/core/mixins/hparams_mixin.py#L33

I'm also ok going for a deprecation, but I should create an issue so we can discuss it there. Nonetheless, I think it's better to go with this documentation change regardless.

**Review by carmocca — APPROVED** [review]

File: docs/source/common/hyperparameters.rst (hunk 3 above)
Comment (carmocca) [reply docs/source/common/hyperparameters.rst:169]: Sure

Comment (awaelchli) [reply docs/source/common/hyperparameters.rst:169]: Created #10375 for everyone to see :)

File: docs/source/common/hyperparameters.rst [hunk 4]
```diff
@@ -113,78 +113,75 @@ LightningModule hyperparameters
...
-                super().__init__()
-                self.layer_1_dim = layer_1_dim
-                self.loss_fx = loss_fx
+Excluding hyperparameters
+"""""""""""""""""""""""""
 
-                # call this to save (layer_1_dim=128) to the checkpoint
-                self.save_hyperparameters("layer_1_dim")
+By default, every parameter of the ``__init__`` method will be considered a hyperparameter to the LightningModule.
+However, sometimes some parameters need to be excluded from saving, for example when they are not serializable.
```
Comment (tchaton) [inline docs/source/common/hyperparameters.rst:148]:
```suggestion
However, sometimes some parameters need to be excluded from saving, for example when they are not serializable.
Those parameters should be provided back when reloading the LightningModule.
```

**Review by tchaton — APPROVED** [review]
LGTM! !

File: docs/source/common/hyperparameters.rst [hunk 5]
```diff
@@ -113,78 +113,75 @@ LightningModule hyperparameters
 Often times we train many versions of a model. You might share that model or come back to it a few months later
 at which point it is very useful to know how that model was trained (i.e.: what learning rate, neural network, etc...).
 
-Lightning has a few ways of saving that information for you in checkpoints and yaml files. The goal here is to
+Lightning has a standardized way of saving that information for you in checkpoints and YAML files. The goal here is to
```
Comment (rohitgr7) [inline docs/source/common/hyperparameters.rst:116]:
```suggestion
Lightning has a standardized way of saving that information for you in checkpoints and YAML files. The goal here is to
```

File: docs/source/common/hyperparameters.rst [hunk 6]
```diff
@@ -113,78 +113,75 @@ LightningModule hyperparameters
...
+            self.layer_1_dim = layer_1_dim
+            self.loss_fx = loss_fx
 
+            # call this to save only (layer_1_dim=128) to the checkpoint
+            self.save_hyperparameters("layer_1_dim")
 
-3.  You can also convert full objects such as ``dict`` or ``Namespace`` to ``hparams`` so they get saved to the
-    checkpoint.
+            # equivalent
+            self.save_hyperparameters(exclude=["loss_fx", "generator_network"])
```
Comment (rohitgr7) [inline docs/source/common/hyperparameters.rst:163]:
```suggestion
            self.save_hyperparameters(ignore=["loss_fx", "generator_network"])
```

**Review by rohitgr7 — APPROVED** [review]
